// preact-cli/lib/lib/webpack/webpack-server-config.js
'use strict';

exports.__esModule = true;

var _webpack = require('@webpack-blocks/webpack2');

var _path = require('path');

var _webpackBaseConfig = require('./webpack-base-config');

var _webpackBaseConfig2 = _interopRequireDefault(_webpackBaseConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (env) {
	var _helpers = (0, _webpackBaseConfig.helpers)(env),
	    cwd = _helpers.cwd;

	return _webpack.createConfig.vanilla([(0, _webpackBaseConfig2.default)(env), (0, _webpack.entryPoint)((0, _path.resolve)(env.cwd, env.src || 'src', 'index.js')), (0, _webpack.setOutput)({
		path: (0, _path.resolve)(cwd, env.dest || 'build', 'ssr-build'),
		publicPath: '/',
		filename: 'ssr-bundle.js',
		chunkFilename: '[name].chunk.[chunkhash:5].js',
		libraryTarget: 'commonjs2'
	}), (0, _webpack.customConfig)({
		target: 'node'
	}), (0, _webpack.customConfig)({
		resolveLoader: {
			alias: {
				'async': (0, _path.resolve)(__dirname, './dummy-loader')
			}
		}
	})]);
};
