// preact-cli/lib/lib/webpack/webpack-base-config.js
'use strict';

exports.__esModule = true;
exports.exists = exists;
exports.helpers = helpers;

var _path = require('path');

var _fs = require('fs');

var _webpack = require('@webpack-blocks/webpack2');

var _extractTextWebpackPlugin = require('extract-text-webpack-plugin');

var _extractTextWebpackPlugin2 = _interopRequireDefault(_extractTextWebpackPlugin);

var _autoprefixer = require('autoprefixer');

var _autoprefixer2 = _interopRequireDefault(_autoprefixer);

var _progressBarWebpackPlugin = require('progress-bar-webpack-plugin');

var _progressBarWebpackPlugin2 = _interopRequireDefault(_progressBarWebpackPlugin);

var _webpackPluginReplace = require('webpack-plugin-replace');

var _webpackPluginReplace2 = _interopRequireDefault(_webpackPluginReplace);

var _requireRelative = require('require-relative');

var _requireRelative2 = _interopRequireDefault(_requireRelative);

var _babelConfig = require('../babel-config');

var _babelConfig2 = _interopRequireDefault(_babelConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function exists(file) {
	try {
		if ((0, _fs.statSync)(file)) return true;
	} catch (e) {}
	return false;
}

function readJson(file) {
	if (file in readJson.cache) return readJson.cache[file];
	var ret = void 0;
	try {
		ret = JSON.parse((0, _fs.readFileSync)(file));
	} catch (e) {}
	return readJson.cache[file] = ret;
}
readJson.cache = {};

function resolveDep(dep, cwd) {
	try {
		return _requireRelative2.default.resolve(dep, cwd || process.cwd());
	} catch (e) {}
	try {
		return require.resolve(dep);
	} catch (e) {}
	return dep;
}

exports.default = function (env) {
	var _helpers = helpers(env),
	    isProd = _helpers.isProd,
	    cwd = _helpers.cwd,
	    src = _helpers.src;

	if (!exists(src('.'))) {
		env.src = '.';
	}

	env.manifest = readJson(src('manifest.json')) || {};
	env.pkg = readJson((0, _path.resolve)(cwd, 'package.json')) || {};

	var babelrc = readJson((0, _path.resolve)(cwd, '.babelrc')) || {};
	var browsers = env.pkg.browserslist || ['> 1%', 'last 2 versions', 'IE >= 9'];

	return (0, _webpack.group)([(0, _webpack.setContext)(src('.')), (0, _webpack.customConfig)({
		resolve: {
			modules: ['node_modules', (0, _path.resolve)(__dirname, '../../../node_modules')],
			extensions: ['.js', '.jsx', '.ts', '.tsx', '.json', '.less', '.scss', '.sass', '.css'],
			alias: {
				'preact-cli-entrypoint': src('index.js'),
				style: src('style'),
				preact$: resolveDep(isProd ? 'preact/dist/preact.min.js' : 'preact', env.cwd),

				react: 'preact-compat',
				'react-dom': 'preact-compat',
				'create-react-class': 'preact-compat/lib/create-react-class',
				'react-addons-css-transition-group': 'preact-css-transition-group'
			}
		},
		resolveLoader: {
			modules: [(0, _path.resolve)(__dirname, '../../../node_modules'), (0, _path.resolve)(cwd, 'node_modules')]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				enforce: 'pre',
				test: /\.jsx?$/,
				loader: 'babel-loader',
				options: Object.assign((0, _babelConfig2.default)(env, { browsers }), babelrc)
			}]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				enforce: 'pre',
				test: /\.less$/,
				use: [{
					loader: (0, _path.resolve)(__dirname, './npm-install-loader'),
					options: {
						modules: ['less', 'less-loader'],
						save: true
					}
				}, {
					loader: 'less-loader',
					options: { sourceMap: true }
				}]
			}, {
				enforce: 'pre',
				test: /\.s[ac]ss$/,
				use: [{
					loader: (0, _path.resolve)(__dirname, './npm-install-loader'),
					options: {
						modules: ['node-sass', 'sass-loader'],
						save: true
					}
				}, {
					loader: 'sass-loader',
					options: { sourceMap: true }
				}]
			}, {
				test: /\.(css|less|s[ac]ss)$/,
				include: [src('components'), src('routes')],
				loader: _extractTextWebpackPlugin2.default.extract({
					fallback: 'style-loader',
					use: [`css-loader?modules&localIdentName=[local]__[hash:base64:5]&importLoaders=1&sourceMap=${isProd}`, `postcss-loader`]
				})
			}, {
				test: /\.(css|less|s[ac]ss)$/,
				exclude: [src('components'), src('routes')],
				loader: _extractTextWebpackPlugin2.default.extract({
					fallback: 'style-loader',
					use: [`css-loader?sourceMap=${isProd}`, `postcss-loader`]
				})
			}]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				test: /\.json$/,
				loader: 'json-loader'
			}, {
				test: /\.(xml|html|txt|md)$/,
				loader: 'raw-loader'
			}, {
				test: /\.(svg|woff2?|ttf|eot|jpe?g|png|gif)(\?.*)?$/i,
				loader: isProd ? 'file-loader' : 'url-loader'
			}]
		}
	}), (0, _webpack.addPlugins)([new _webpack.webpack.LoaderOptionsPlugin({
		options: {
			postcss: function postcss() {
				return [(0, _autoprefixer2.default)({ browsers })];
			},
			context: (0, _path.resolve)(cwd, env.src || 'src')
		}
	})]), (0, _webpack.defineConstants)({
		'process.env.NODE_ENV': isProd ? 'production' : 'development'
	}), (0, _webpack.setDevTool)(isProd ? 'source-map' : 'cheap-module-eval-source-map'), (0, _webpack.customConfig)({
		node: {
			console: false,
			process: false,
			Buffer: false,
			__filename: false,
			__dirname: false,
			setImmediate: false
		}
	}), (0, _webpack.addPlugins)([new _extractTextWebpackPlugin2.default({
		filename: 'style.css',
		disable: !isProd,
		allChunks: true
	})]), isProd ? production() : development(), (0, _webpack.addPlugins)([new _webpack.webpack.NoEmitOnErrorsPlugin(), new _progressBarWebpackPlugin2.default({
		format: '\u001b[90m\u001b[44mBuild\u001b[49m\u001b[39m [:bar] \u001b[32m\u001b[1m:percent\u001b[22m\u001b[39m (:elapseds) \u001b[2m:msg\u001b[22m',
		renderThrottle: 100,
		summary: false,
		clear: true
	}), new _webpack.webpack.optimize.CommonsChunkPlugin({
		async: false,
		children: true,
		minChunks: 3
	})])].filter(Boolean));
};

var development = function development() {
	return (0, _webpack.group)([]);
};

var production = function production() {
	return (0, _webpack.addPlugins)([new _webpack.webpack.HashedModuleIdsPlugin(), new _webpack.webpack.LoaderOptionsPlugin({
		minimize: true
	}), new _webpackPluginReplace2.default({
		include: /babel-helper$/,
		patterns: [{
			regex: /throw\s+(new\s+)?(Type|Reference)?Error\s*\(/g,
			value: function value(s) {
				return `return;${Array(s.length - 7).join(' ')}(`;
			}
		}]
	}), new _webpack.webpack.optimize.UglifyJsPlugin({
		output: {
			comments: false
		},
		mangle: true,
		sourceMap: true,
		compress: {
			properties: true,
			keep_fargs: false,
			pure_getters: true,
			collapse_vars: true,
			warnings: false,
			screw_ie8: true,
			sequences: true,
			dead_code: true,
			drop_debugger: true,
			comparisons: true,
			conditionals: true,
			evaluate: true,
			booleans: true,
			loops: true,
			unused: true,
			hoist_funs: true,
			if_return: true,
			join_vars: true,
			cascade: true,
			drop_console: false,
			pure_funcs: ['classCallCheck', '_classCallCheck', '_possibleConstructorReturn', 'Object.freeze', 'invariant', 'warning']
		}
	})]);
};

function helpers(env) {
	return {
		isProd: env && env.production,
		cwd: env.cwd = (0, _path.resolve)(env.cwd || process.cwd()),
		src: function src(dir) {
			return (0, _path.resolve)(env.cwd, env.src || 'src', dir);
		}
	};
}
