// preact-cli/lib/lib/webpack-config.js
'use strict';

exports.__esModule = true;

var _path = require('path');

var _fs = require('fs');

var _minimatch = require('minimatch');

var _webpack = require('@webpack-blocks/webpack2');

var _devServer = require('@webpack-blocks/dev-server2');

var _devServer2 = _interopRequireDefault(_devServer);

var _extractTextWebpackPlugin = require('extract-text-webpack-plugin');

var _extractTextWebpackPlugin2 = _interopRequireDefault(_extractTextWebpackPlugin);

var _autoprefixer = require('autoprefixer');

var _autoprefixer2 = _interopRequireDefault(_autoprefixer);

var _htmlWebpackPlugin = require('html-webpack-plugin');

var _htmlWebpackPlugin2 = _interopRequireDefault(_htmlWebpackPlugin);

var _scriptExtHtmlWebpackPlugin = require('script-ext-html-webpack-plugin');

var _scriptExtHtmlWebpackPlugin2 = _interopRequireDefault(_scriptExtHtmlWebpackPlugin);

var _htmlWebpackExcludeAssetsPlugin = require('html-webpack-exclude-assets-plugin');

var _htmlWebpackExcludeAssetsPlugin2 = _interopRequireDefault(_htmlWebpackExcludeAssetsPlugin);

var _progressBarWebpackPlugin = require('progress-bar-webpack-plugin');

var _progressBarWebpackPlugin2 = _interopRequireDefault(_progressBarWebpackPlugin);

var _copyWebpackPlugin = require('copy-webpack-plugin');

var _copyWebpackPlugin2 = _interopRequireDefault(_copyWebpackPlugin);

var _webpackPluginReplace = require('webpack-plugin-replace');

var _webpackPluginReplace2 = _interopRequireDefault(_webpackPluginReplace);

var _swPrecacheWebpackPlugin = require('sw-precache-webpack-plugin');

var _swPrecacheWebpackPlugin2 = _interopRequireDefault(_swPrecacheWebpackPlugin);

var _requireRelative = require('require-relative');

var _requireRelative2 = _interopRequireDefault(_requireRelative);

var _prerender = require('./prerender');

var _prerender2 = _interopRequireDefault(_prerender);

var _pushManifest = require('./push-manifest');

var _pushManifest2 = _interopRequireDefault(_pushManifest);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function exists(file) {
	try {
		if ((0, _fs.statSync)(file)) return true;
	} catch (e) {}
	return false;
}

function readJson(file) {
	if (file in readJson.cache) return readJson.cache[file];
	var ret = void 0;
	try {
		ret = JSON.parse((0, _fs.readFileSync)(file));
	} catch (e) {}
	return readJson.cache[file] = ret;
}
readJson.cache = {};

function resolveDep(dep, cwd) {
	try {
		return _requireRelative2.default.resolve(dep, cwd || process.cwd());
	} catch (e) {}
	try {
		return require.resolve(dep);
	} catch (e) {}
	return dep;
}

exports.default = function (env) {
	var isProd = env && env.production;
	var cwd = env.cwd = (0, _path.resolve)(env.cwd || process.cwd());
	var src = function src(dir) {
		return (0, _path.resolve)(env.cwd, env.src || 'src', dir);
	};

	if (!exists(src('.'))) {
		env.src = '.';
	}

	env.manifest = readJson(src('manifest.json')) || {};
	env.pkg = readJson((0, _path.resolve)(cwd, 'package.json')) || {};

	var browsers = env.pkg.browserslist || ['> 1%', 'last 2 versions', 'IE >= 9'];

	return _webpack.createConfig.vanilla([(0, _webpack.setContext)(src('.')), (0, _webpack.entryPoint)({
		'bundle': (0, _path.resolve)(__dirname, './entry'),
		'polyfills': (0, _path.resolve)(__dirname, './polyfills')
	}), (0, _webpack.setOutput)({
		path: (0, _path.resolve)(cwd, env.dest || 'build'),
		publicPath: '/',
		filename: '[name].js',
		chunkFilename: '[name].chunk.[chunkhash:5].js'
	}), (0, _webpack.customConfig)({
		resolve: {
			modules: ['node_modules', (0, _path.resolve)(__dirname, '../../node_modules')],
			extensions: ['.js', '.jsx', '.ts', '.tsx', '.json', '.less', '.scss', '.sass', '.css'],
			alias: {
				'preact-cli-entrypoint': src('index.js'),
				style: src('style'),
				preact$: resolveDep(isProd ? 'preact/dist/preact.min.js' : 'preact', env.cwd),

				react: 'preact-compat',
				'react-dom': 'preact-compat',
				'create-react-class': 'preact-compat/lib/create-react-class',
				'react-addons-css-transition-group': 'preact-css-transition-group'
			}
		},
		resolveLoader: {
			alias: {
				'async': (0, _path.resolve)(__dirname, './async-component-loader')
			},
			modules: [(0, _path.resolve)(__dirname, '../../node_modules'), (0, _path.resolve)(cwd, 'node_modules')]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				enforce: 'pre',
				test: /\.jsx?$/,
				loader: 'babel-loader',
				options: {
					babelrc: true,
					presets: [[(0, _path.resolve)(__dirname, './babel-config'), { browsers }]]
				}
			}]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				test: /\.jsx?$/,
				include: [(0, _minimatch.filter)(src('routes') + '/{*.js,*/index.js}'), (0, _minimatch.filter)(src('components') + '/{routes,async}/{*.js,*/index.js}')],
				loader: (0, _path.resolve)(__dirname, './async-component-loader'),
				options: {
					name(filename) {
						var relative = filename.replace(src('.'), '');
						var isRoute = filename.indexOf('/routes/') >= 0;

						return isRoute ? 'route-' + relative.replace(/(^\/(routes|components\/(routes|async))\/|(\/index)?\.js$)/g, '') : false;
					},
					formatName(filename) {
						var relative = filename.replace(src('.'), '');

						return relative.replace(/(^\/(routes|components\/(routes|async))\/|(\/index)?\.js$)/g, '');
					}
				}
			}]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				enforce: 'pre',
				test: /\.less$/,
				use: [{
					loader: (0, _path.resolve)(__dirname, './npm-install-loader'),
					options: {
						modules: ['less', 'less-loader'],
						save: true
					}
				}, {
					loader: 'less-loader',
					options: { sourceMap: true }
				}]
			}, {
				enforce: 'pre',
				test: /\.s[ac]ss$/,
				use: [{
					loader: (0, _path.resolve)(__dirname, './npm-install-loader'),
					options: {
						modules: ['node-sass', 'sass-loader'],
						save: true
					}
				}, {
					loader: 'sass-loader',
					options: { sourceMap: true }
				}]
			}, {
				test: /\.(css|less|s[ac]ss)$/,
				include: [src('components'), src('routes')],
				loader: _extractTextWebpackPlugin2.default.extract({
					fallback: 'style-loader',
					use: [`css-loader?modules&localIdentName=[local]__[hash:base64:5]&importLoaders=1&sourceMap=${isProd}`, `postcss-loader`]
				})
			}, {
				test: /\.(css|less|s[ac]ss)$/,
				exclude: [src('components'), src('routes')],
				loader: _extractTextWebpackPlugin2.default.extract({
					fallback: 'style-loader',
					use: [`css-loader?sourceMap=${isProd}`, `postcss-loader`]
				})
			}]
		}
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				test: /\.json$/,
				loader: 'json-loader'
			}, {
				test: /\.(xml|html|txt|md)$/,
				loader: 'raw-loader'
			}, {
				test: /\.(svg|woff2?|ttf|eot|jpe?g|png|gif)(\?.*)?$/i,
				loader: isProd ? 'file-loader' : 'url-loader'
			}]
		}
	}), (0, _webpack.addPlugins)([new _webpack.webpack.LoaderOptionsPlugin({
		options: {
			postcss: function postcss() {
				return [(0, _autoprefixer2.default)({ browsers })];
			},
			context: (0, _path.resolve)(cwd, env.src || 'src')
		}
	})]), (0, _webpack.defineConstants)({
		'process.env.NODE_ENV': isProd ? 'production' : 'development'
	}), isProd && (0, _webpack.performance)(Object.assign({
		maxAssetSize: 200 * 1000,
		maxEntrypointSize: 200 * 1000,
		hints: 'warning'
	}, env.pkg.performance || {})), (0, _webpack.setDevTool)(isProd ? 'source-map' : 'cheap-module-eval-source-map'), (0, _webpack.customConfig)({
		node: {
			console: false,
			process: false,
			Buffer: false,
			__filename: false,
			__dirname: false,
			setImmediate: false
		}
	}), (0, _webpack.addPlugins)([new _copyWebpackPlugin2.default([].concat(exists(src('manifest.json')) ? [{ from: 'manifest.json' }] : [{
		from: (0, _path.resolve)(__dirname, '../resources/manifest.json'),
		to: 'manifest.json'
	}, {
		from: (0, _path.resolve)(__dirname, '../resources/icon.png'),
		to: 'assets/icon.png'
	}], [exists(src('assets')) && {
		from: 'assets',
		to: 'assets'
	}]).filter(Boolean))]), (0, _webpack.addPlugins)([new _extractTextWebpackPlugin2.default({
		filename: 'style.css',
		disable: !isProd,
		allChunks: true
	})]), htmlPlugin(env), isProd ? production(env) : development(env), (0, _webpack.addPlugins)([new _webpack.webpack.NoEmitOnErrorsPlugin(), new _progressBarWebpackPlugin2.default({
		format: '\u001b[90m\u001b[44mBuild\u001b[49m\u001b[39m [:bar] \u001b[32m\u001b[1m:percent\u001b[22m\u001b[39m (:elapseds) \u001b[2m:msg\u001b[22m',
		renderThrottle: 100,
		summary: false,
		clear: true
	}), new _webpack.webpack.optimize.CommonsChunkPlugin({
		async: false,
		children: true,
		minChunks: 3
	}), new _pushManifest2.default()])].filter(Boolean));
};

var development = function development(config) {
	var port = process.env.PORT || config.port || 8080,
	    host = process.env.HOST || config.host || '0.0.0.0',
	    origin = `${config.https === true ? 'https' : 'http'}://${host}:${port}/`;

	return (0, _webpack.group)([(0, _webpack.addPlugins)([new _webpack.webpack.NamedModulesPlugin()]), (0, _devServer2.default)({
		port,
		host,
		inline: true,
		hot: true,
		https: config.https,
		compress: true,
		publicPath: '/',
		contentBase: (0, _path.resolve)(config.cwd, config.src || './src'),

		disableHostCheck: true,
		historyApiFallback: true,
		quiet: true,
		clientLogLevel: 'none',
		overlay: false,
		stats: 'minimal',
		watchOptions: {
			ignored: [(0, _path.resolve)(config.cwd, 'build'), (0, _path.resolve)(config.cwd, 'node_modules')]
		}
	}, [`webpack-dev-server/client?${origin}`, `webpack/hot/dev-server?${origin}`])]);
};

var production = function production(config) {
	return (0, _webpack.addPlugins)([new _webpack.webpack.HashedModuleIdsPlugin(), new _webpack.webpack.LoaderOptionsPlugin({
		minimize: true
	}), new _webpackPluginReplace2.default({
		include: /babel-helper$/,
		patterns: [{
			regex: /throw\s+(new\s+)?(Type|Reference)?Error\s*\(/g,
			value: function value(s) {
				return `return;${Array(s.length - 7).join(' ')}(`;
			}
		}]
	}), new _webpack.webpack.optimize.UglifyJsPlugin({
		output: {
			comments: false
		},
		mangle: true,
		sourceMap: true,
		compress: {
			properties: true,
			keep_fargs: false,
			pure_getters: true,
			collapse_vars: true,
			warnings: false,
			screw_ie8: true,
			sequences: true,
			dead_code: true,
			drop_debugger: true,
			comparisons: true,
			conditionals: true,
			evaluate: true,
			booleans: true,
			loops: true,
			unused: true,
			hoist_funs: true,
			if_return: true,
			join_vars: true,
			cascade: true,
			drop_console: false,
			pure_funcs: ['classCallCheck', '_classCallCheck', '_possibleConstructorReturn', 'Object.freeze', 'invariant', 'warning']
		}
	}), new _swPrecacheWebpackPlugin2.default({
		filename: 'sw.js',
		navigateFallback: 'index.html',
		navigateFallbackWhitelist: [/^(?!\/__).*/],
		minify: true,
		stripPrefix: config.cwd,
		staticFileGlobsIgnorePatterns: [/polyfills\.js$/, /\.map$/, /push-manifest\.json$/]
	})]);
};

var htmlPlugin = function htmlPlugin(config) {
	return (0, _webpack.addPlugins)([new _htmlWebpackPlugin2.default({
		filename: 'index.html',
		template: `!!ejs-loader!${config.template || (0, _path.resolve)(__dirname, '../resources/template.html')}`,
		minify: config.production && {
			collapseWhitespace: true,
			removeScriptTypeAttributes: true,
			removeRedundantAttributes: true,
			removeStyleLinkTypeAttributes: true,
			removeComments: true
		},
		favicon: exists((0, _path.resolve)(config.src, 'assets/favicon.ico')) ? 'assets/favicon.ico' : (0, _path.resolve)(__dirname, '../resources/favicon.ico'),
		manifest: config.manifest,
		inject: true,
		compile: true,
		preload: config.preload === true,
		title: config.title || config.manifest.name || config.manifest.short_name || (config.pkg.name || '').replace(/^@[a-z]\//, '') || 'Preact App',
		excludeAssets: [/(bundle|polyfills)(\..*)?\.js$/],
		config,
		ssr(params) {
			return config.prerender ? (0, _prerender2.default)(config, params) : '';
		}
	}), new _htmlWebpackExcludeAssetsPlugin2.default(), new _scriptExtHtmlWebpackPlugin2.default({
		defaultAttribute: 'defer'
	})]);
};
